src/poll.iops.cpp

```diff
@@ -62,92 +68,117 @@ struct felspar::io::poll_warden::write_some_completion :
-    int fd;
+    : completion<socket_descriptor>{s, t, loc}, fd{f} {}
+    socket_descriptor fd;
     void cancel_iop() override { std::erase(self->requests[fd].reads, this); }
     felspar::coro::coroutine_handle<> try_or_resume() override {
-#ifdef FELSPAR_HAS_ACCEPT4
-        auto const r = ::accept4(fd, nullptr, nullptr, SOCK_NONBLOCK);
-        if (r >= 0) {
+#ifdef FELSPAR_WINSOCK2
+        if (auto const r = ::accept(fd, nullptr, nullptr);
+            r != INVALID_SOCKET) {
+#elif defined(FELSPAR_HAS_ACCEPT4)
+        if (auto const r = ::accept4(fd, nullptr, nullptr, SOCK_NONBLOCK);
+            r >= 0) {
 #else
-        auto const r = ::accept(fd, nullptr, nullptr);
-        if (r >= 0) {
+        if (auto const r = ::accept(fd, nullptr, nullptr); r >= 0) {
             posix::set_non_blocking(r, loc);
 #endif
             result = r;
             return cancel_timeout_then_resume();
-        } else if (errno == EWOULDBLOCK or errno == EAGAIN) {
+        } else if (auto const error = get_error(); would_block(error)) {
             self->requests[fd].reads.push_back(this);
             return felspar::coro::noop_coroutine();
-        } else if (errno == EBADF) {
+        } else if (bad_fd(error)) {
             result = r;
             return cancel_timeout_then_resume();
         } else {
-            result = {{errno, std::system_category()}, "accept"};
+            result = {{error, std::system_category()}, "accept"};
             return cancel_timeout_then_resume();
         }
     }
 };
-felspar::io::iop<int> felspar::io::poll_warden::do_accept(
-        int fd,
-        std::optional<std::chrono::nanoseconds> timeout,
-        felspar::source_location const &loc) {
+felspar::io::iop<felspar::io::socket_descriptor>
+        felspar::io::poll_warden::do_accept(
+                socket_descriptor fd,
+                std::optional<std::chrono::nanoseconds> timeout,
+                felspar::source_location const &loc) {
     return {new accept_completion{this, fd, timeout, loc}};
 }
 
 
 struct felspar::io::poll_warden::connect_completion : public completion<void> {
     connect_completion(
             poll_warden *s,
-            int f,
+            socket_descriptor f,
             sockaddr const *a,
             socklen_t l,
             std::optional<std::chrono::nanoseconds> t,
             felspar::source_location const &loc)
     : completion<void>{s, t, loc}, fd{f}, addr{a}, addrlen{l} {}
-    int fd;
+    socket_descriptor fd;
     sockaddr const *addr;
     socklen_t addrlen;
     void cancel_iop() override { std::erase(self->requests[fd].writes, this); }
     felspar::coro::coroutine_handle<>
             await_suspend(felspar::coro::coroutine_handle<> h) override {
         handle = h;
+#ifdef FELSPAR_WINSOCK2
+        if (auto err = ::connect(fd, addr, addrlen); err != SOCKET_ERROR) {
+            return handle;
+        } else if (auto const wsae = WSAGetLastError();
+                   wsae == WSAEWOULDBLOCK) {
```

Comment:
Not so easy to scroll back on my phone but didn't you introduce a "would block" helper earlier?

Reply:
There is, but this is Windows code anyway. Still, I guess no harm in using it